### PR TITLE
Add msgpack serialization support

### PR DESCRIPTION
**Pull request by goodjava**
Add msgpack serialization support

支持msgpack方式的序列化协议
499E2DCB-5C47-4EED-8180-E57562468B45

### CODE REVIEW

File: dubbo-serialization/dubbo-serialization-msgpack/src/main/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectInput.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import org.apache.dubbo.common.serialize.ObjectInput;
+import org.apache.dubbo.common.utils.PojoUtils;
+
+import org.msgpack.jackson.dataformat.MessagePackFactory;
+
+import java.io.*;
```
Comment (AlbumenJ) [inline dubbo-serialization/dubbo-serialization-msgpack/src/main/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectInput.java:28]: Please do not use comma import

File: dubbo-serialization/dubbo-serialization-msgpack/src/main/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectOutput.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+package org.apache.dubbo.common.serialize.msgpack;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import org.apache.dubbo.common.serialize.ObjectOutput;
+
+import org.msgpack.jackson.dataformat.MessagePackFactory;
+
+import java.io.*;
```
Comment (AlbumenJ) [inline dubbo-serialization/dubbo-serialization-msgpack/src/main/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectOutput.java:26]: also for this

File: dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/BizException.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+
```
Comment (AlbumenJ) [inline dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/BizException.java:1]: Please add Apache License header for this file

File: dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectOutputTest.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+import static org.hamcrest.CoreMatchers.*;
```
Comment (AlbumenJ) [inline dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectOutputTest.java:28]: also for this file

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectOutputTest.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    public void testWriteByte() throws IOException {
+        this.msgpackObjectOutput.writeByte((byte) 123);
+        this.flushToInput();
+
+        assertThat(this.msgpackObjectInput.readByte(), is((byte) 123));
+    }
+
+    @Test
+    public void testWriteObject() throws IOException, ClassNotFoundException {
+        Image image = new Image("http://dubbo.apache.org/img/dubbo_white.png", "logo", 300, 480, Image.Size.SMALL);
```
Comment (AlbumenJ) [inline dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectOutputTest.java:147]: Please save this file to test resources. Try to reduce network dependence for unit test cases.

File: dubbo-serialization/dubbo-serialization-msgpack/src/main/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectInput.java (hunk 1 above)
Comment (goodjava) [reply dubbo-serialization/dubbo-serialization-msgpack/src/main/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectInput.java:28]: get

File: dubbo-serialization/dubbo-serialization-msgpack/src/main/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectInput.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+public class MsgpackObjectInput implements ObjectInput {
+
+    private final InputStream in;
+
+    private ObjectMapper om;
+
+    public MsgpackObjectInput(InputStream in) {
+        this.in = in;
+        om = new ObjectMapper(new MessagePackFactory());
+        om.configure(JsonParser.Feature.AUTO_CLOSE_SOURCE, false);
```
Comment (pinxiong) [inline dubbo-serialization/dubbo-serialization-msgpack/src/main/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectInput.java:40]: Why can't we initialize `ObjectMapper` in the same way in both `MsgpackObjectInput` and `MsgpackObjectOutput` class?
```java
    public MsgpackObjectOutput(OutputStream out) {
        this.out = out;
        om = new ObjectMapper(new MessagePackFactory());
    }
```

File: dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectOutputTest.java (hunk 4 above)
Comment (goodjava) [reply dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectOutputTest.java:28]: get

File: dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectOutputTest.java (hunk 5 above)
Comment (goodjava) [reply dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectOutputTest.java:147]: get

File: dubbo-serialization/dubbo-serialization-msgpack/src/main/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectInput.java (hunk 6 above)
Comment (goodjava) [reply dubbo-serialization/dubbo-serialization-msgpack/src/main/java/org/apache/dubbo/common/serialize/msgpack/MsgpackObjectInput.java:40]: get

File: dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/BizException.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.dubbo.common.serialize.msgpack;
+
+public class BizException extends RuntimeException {
```
Comment (horizonzy) [inline dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/BizException.java:19]: change it to static class inner In Test class.

**Review by xiaoheng1 — COMMENTED** [review]
agree with horizonzy. and also Image need remove to Test class.

File: dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/BizException.java (hunk 7 above)
Comment (goodjava) [reply dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/BizException.java:19]: get

**Comment by goodjava** [comment]
> agree with horizonzy. and also Image need remove to Test class.

get

File: dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/BizException.java (hunk 3 above)
Comment (goodjava) [reply dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/BizException.java]: for test @xiaoheng1

File: dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/BizException.java (hunk 7 above)
Comment (endeavor66) [reply dubbo-serialization/dubbo-serialization-msgpack/src/test/java/org/apache/dubbo/common/serialize/msgpack/BizException.java]: it's ok

**Review by endeavor66 — COMMENTED** [review]
a test comment

**Pull request merged**